### PR TITLE
Updated content in various nodes in NodeJS  -Why nodejs, Installing N…

### PR DESCRIPTION
**Pull request by CodeWAdi**
…odeJS , CommonJS vs ESM , Global keyword,npm workspaces , creating packages , __dirname , __filename

Updated various sub groups with information in NODEJS

Merge the PR if you think its correct and add HACKTOBERFEST-ACCEPTED if you think this helps. PLEASE

### CODE REVIEW

File: content/roadmaps/107-nodejs/content/100-nodejs-introduction/101-why-nodejs.md [hunk 1]
```diff
@@ -1 +1,4 @@
-# Why nodejs
\ No newline at end of file
+# Why nodejs
+
+Node.js is an open source, a server-side script which runs on the top of Google’s open-source scripting engine V8. Node.js is fast, lightweight and efficient. It uses the asynchronous mode of operation, event-driven Input/Output rather than using the traditional threads or separate threads for each process. Node.js was originally written by Ryan Dahl in the year 2009. It is a cross-platform Javascript run-time environment that executes Javascript code outside of a browser. Node.js uses javascript for creating node applications or we can use any other language that ultimately compiles to javascript (like typescript). The javascript is written in the same way as we’d use in any client-side application. However, we need to set up the node development environment.
+
```
Comment (kamranahmedse) [inline content/roadmaps/107-nodejs/content/100-nodejs-introduction/101-why-nodejs.md:4]:
```suggestion
# Why Node.js

* Node.js is OpenSource
* Guarantees ease of scaling applications vertically by adding new resources to the existing node and horizontally with the addition of new nodes
* It facilitates building separate components that easily complement larger applications, which means saving money at the initial development stages and potentially shortening time-to-market, reusable code (some of the code can be used both on the front and backend)
* Tested on production (a lot of [companies use it](https://selleo.com/blog/10-successful-companies-using-nodejs) from the very beginning in 2009)
* offers improved performance thanks to Chrome’s V8 engine
* using JavaScript on both backend and frontend makes Node a great choice to improve speed and simplicity of implementation
```

File: content/roadmaps/107-nodejs/content/100-nodejs-introduction/101-why-nodejs.md [hunk 2]
```diff
@@ -1 +1,8 @@
-# Why nodejs
\ No newline at end of file
+# Why Node.js
+
+* Node.js is OpenSource
+* Guarantees ease of scaling applications vertically by adding new resources to the existing node and horizontally with the addition of new nodes
+* It facilitates building separate components that easily complement larger applications, which means saving money at the initial development stages and potentially shortening time-to-market, reusable code (some of the code can be used both on the front and backend)
+* Tested on production (a lot of [companies use it](https://selleo.com/blog/10-successful-companies-using-nodejs) from the very beginning in 2009)
+* offers improved performance thanks to Chrome’s V8 engine
+* using JavaScript on both backend and frontend makes Node a great choice to improve speed and simplicity of implementation
```
Comment (kamranahmedse) [inline content/roadmaps/107-nodejs/content/100-nodejs-introduction/101-why-nodejs.md:8]:
```suggestion
* using JavaScript on both backend and frontend makes Node a great choice to improve speed and simplicity of implementation

<ResourceGroupTitle>Free Content</ResourceGroupTitle>
<BadgeLink colorScheme='yellow' badgeText='Read' href='https://nodejs.dev/en/learn/'>Learn Node.js</BadgeLink>
<BadgeLink colorScheme='yellow' badgeText='Read' href='https://medium.com/selleo/why-choose-node-js-b0091ad6c3fc'>Why Choose Node.js?</BadgeLink>
<BadgeLink colorScheme='yellow' badgeText='Read' href='https://www.bitovi.com/blog/5-reasons-to-choose-nodejs'>5 Reasons to Choose Node.js</BadgeLink>
```

File: content/roadmaps/107-nodejs/content/100-nodejs-introduction/104-running-nodejs-code.md [hunk 3]
```diff
@@ -1 +1,58 @@
...
+The final step in node.js installed is the updation of your local npm version(if required) – the package manager that comes bundled with Node.js.
+
+You can run the following command, to quickly update the npm
+
+npm install npm –global // Updates the ‘CLI’ client
+
+<ResourceGroupTitle>Free Content</ResourceGroupTitle>
+<BadgeLink colorScheme='blue' badgeText='Read' href='https://nodejs.org/en/about/'>Official Website</BadgeLink>
+<BadgeLink colorScheme='yellow' badgeText='Read' href='https://www.w3schools.com/nodejs/nodejs_intro.asp'>Node.JS Introduction</BadgeLink>
+<BadgeLink colorScheme='green' badgeText='Read' href='https://nodejs.org/en/download/'>Official Installation Link</BadgeLink>v
```
Comment (kamranahmedse) [inline content/roadmaps/107-nodejs/content/100-nodejs-introduction/104-running-nodejs-code.md:58]:
```suggestion
# Running Node.js Code

The usual way to run a Node.js program is to run the globally available `node` command (once you install Node.js) and pass the name of the file you want to execute.

<ResourceGroupTitle>Free Content</ResourceGroupTitle>
<BadgeLink colorScheme='yellow' badgeText='Read' href='https://nodejs.dev/en/learn/run-nodejs-scripts-from-the-command-line/'>Run Node.js from Command Line</BadgeLink>
```

File: content/roadmaps/107-nodejs/content/101-nodejs-modules/100-commonjs-vs-esm.md [hunk 4]
```diff
@@ -1 +1,9 @@
...
\ No newline at end of file
+# Commonjs vs esm
+
+The CommonJS module system, on the other hand, is built into Node.js. Prior to the introduction of the ES module in Node.js, CommonJS was the standard for Node.js modules. As a result, there are plenty of Node.js libraries and modules written with CommonJS.
+
+The ES module format was created to standardize the JavaScript module system. It has become the standard format for encapsulating JavaScript code for reuse.
+
+<ResourceGroupTitle>Free Content</ResourceGroupTitle>
+<BadgeLink colorScheme='blue' badgeText='Read' href='https://nodejs.org/en/about/'>Official Website</BadgeLink>
+<BadgeLink colorScheme='yellow' badgeText='Read' href='https://reflectoring.io/nodejs-modules-imports/'>CommonJs VS ESM</BadgeLink>
```
Comment (kamranahmedse) [inline content/roadmaps/107-nodejs/content/101-nodejs-modules/100-commonjs-vs-esm.md:9]:
```suggestion
# CommonJS vs ESM

The CommonJS module system, on the other hand, is built into Node.js. Before the introduction of the ES module in Node.js, CommonJS was the standard for Node.js modules. As a result, many Node.js libraries and modules are written with CommonJS.

The ES module format was created to standardize the JavaScript module system. It has become the standard format for encapsulating JavaScript code for reuse.

<ResourceGroupTitle>Free Content</ResourceGroupTitle>
<BadgeLink colorScheme='blue' badgeText='Read' href='https://nodejs.org/en/about/'>Official Website</BadgeLink>
<BadgeLink colorScheme='yellow' badgeText='Read' href='https://reflectoring.io/nodejs-modules-imports/'>CommonJs VS ESM</BadgeLink> 
```

File: content/roadmaps/107-nodejs/content/101-nodejs-modules/102-global-keyword.md [hunk 5]
```diff
@@ -1 +1,24 @@
...
+ 5.URLSearchParams: URLSearchParams API is used to perform read and write operations on the query of a URL. 
+
+ 6.WebAssembly: The global object that acts as a namespace for all W3C WebAssembly related functionality. WebAssembly is a low level Assembly-like language that can be run on modern browsers. 
+
+ 7.__filename: The output throws an error which proves that __filename is not globally defined in node.js. It requires a script to give the desired output as __filename is only defined in scripts.
+
+ <ResourceGroupTitle>Free Content</ResourceGroupTitle>
+<BadgeLink colorScheme='blue' badgeText='Read' href='https://nodejs.org/en/about/'>Official Website</BadgeLink>
+<BadgeLink colorScheme='yellow' badgeText='Read' href='https://www.geeksforgeeks.org/node-js-global-objects/'>Node.JS Global</BadgeLink>
+<BadgeLink colorScheme='green' badgeText='Read' href='https://nodejs.org/api/globals.html'>About Global</BadgeLink>
```
Comment (kamranahmedse) [inline content/roadmaps/107-nodejs/content/101-nodejs-modules/102-global-keyword.md:24]:
```suggestion
# global Keyword

In browsers, the top-level scope is the global scope. This means that within the browser var something will define a new global variable. In Node.js this is different. The top-level scope is not the global scope; `var something` inside a Node.js module will be local to that module.

<ResourceGroupTitle>Free Content</ResourceGroupTitle>
<BadgeLink colorScheme='yellow' badgeText='Read' href='https://nodejs.org/api/globals.html#global'>global Keyword in Node.js</BadgeLink>
```

File: content/roadmaps/107-nodejs/content/102-nodejs-npm/105-npm-workspaces.md [hunk 6]
```diff
@@ -1 +1,13 @@
...
+
+This set of features makes up for a much more streamlined workflow handling linked packages from the local file system. Automating the linking process as part of npm install and avoiding manually having to use npm link in order to add references to packages that should be symlinked into the current node_modules folder.
+
+We also refer to these packages being auto-symlinked during npm install as a single workspace, meaning it's a nested package within the current local file system that is explicitly defined in the package.json workspaces configuration.
+
+<ResourceGroupTitle>Free Content</ResourceGroupTitle>
+<BadgeLink colorScheme='blue' badgeText='Read' href='https://ruanmartinelli.com/posts/npm-7-workspaces-1'>Getting Started with Npm Workspaces </BadgeLink>
+<BadgeLink colorScheme='yellow' badgeText='Read' href='https://docs.npmjs.com/cli/v8/using-npm/workspaces'>Official | Npm Workspaces</BadgeLink>
+<BadgeLink badgeText='Watch' href='https://www.youtube.com/watch?v=LJaLwlS0mj4'>npm workspaces and overrides</BadgeLink>
+<BadgeLink badgeText='Watch' href='https://www.youtube.com/watch?v=e9ABjLAi27Q'>npm workspaces</BadgeLink>
```
Comment (kamranahmedse) [inline content/roadmaps/107-nodejs/content/102-nodejs-npm/105-npm-workspaces.md:13]:
```suggestion
# npm workspaces

Workspace is a generic term that refers to the set of npm CLI features that support managing multiple packages from your local file system from within a singular top-level root package.

<ResourceGroupTitle>Free Content</ResourceGroupTitle>
<BadgeLink colorScheme='blue' badgeText='Official Docs' href='https://docs.npmjs.com/cli/v8/using-npm/workspaces'>npm workspaces</BadgeLink>
<BadgeLink colorScheme='yellow' badgeText='Read' href='https://ruanmartinelli.com/posts/npm-7-workspaces-1'>Getting Started with Npm Workspaces </BadgeLink>
```

File: content/roadmaps/107-nodejs/content/102-nodejs-npm/106-creating-packages.md [hunk 7]
```diff
@@ -1 +1,16 @@
...
+    For unscoped modules, run npm init
+ 2. Provide responses for the required fields (name and version), as well as the main field:
+    name: The name of your module.
+    version: The initial module version. We recommend following semantic versioning guidelines and starting with 1.0.0.
+
+ <ResourceGroupTitle>Free Content</ResourceGroupTitle>
+ <BadgeLink colorScheme='blue' badgeText='Read' href='https://medium.com/medialesson/create-your-own-npm-package-private-or-public-60edc2c42f68'>Public and Private packages</BadgeLink>
+ <BadgeLink colorScheme='yellow' badgeText='Read' href='https://docs.npmjs.com/creating-node-js-modules'>Official | Creating packages</BadgeLink>
+ <BadgeLink badgeText='Watch' href='https://www.youtube.com/watch?v=J4b_T-qH3BY'>Create and Publish your first packages</BadgeLink>
+ <BadgeLink badgeText='Watch' href='https://www.youtube.com/watch?v=4zzbNac6f6Q'>Create Packages </BadgeLink>
```
Comment (kamranahmedse) [inline content/roadmaps/107-nodejs/content/102-nodejs-npm/106-creating-packages.md:16]:
```suggestion
# Creating Packages

npm packages allow you to bundle some specific functionality into a reusable package which can then be uploaded to some package registry such as npm or GitHub packages and then be installed and reused in projects using npm.

<ResourceGroupTitle>Free Content</ResourceGroupTitle>
<BadgeLink colorScheme='yellow' badgeText='Read' href='https://www.freecodecamp.org/news/how-to-make-a-beautiful-tiny-npm-package-and-publish-it-2881d4307f78/'>How to make a tiny npm package and publish it</BadgeLink>
<BadgeLink colorScheme='yellow' badgeText='Read' href='https://snyk.io/blog/best-practices-create-modern-npm-package/'>Best practices for creating a modern npm package</BadgeLink>
```

File: content/roadmaps/107-nodejs/content/105-nodejs-working-with-files/107-dirname.md [hunk 8]
```diff
@@ -1 +1,11 @@
...
+
+The __dirname in a node script returns the path of the folder where the current JavaScript file resides. __filename and __dirname are used to get the filename and directory name of the currently executing file.
+
+1. Gives absolute path of the directory that contains the currently executing file.	
+2. Returns a pointer to a string i.e. the parent directory of currently executing file.	
+3. Works similar to process.cwd() until the node is run from a directory which is different from the directory where the JavaScript file is stored.
+
+<ResourceGroupTitle>Free Content</ResourceGroupTitle>
+<BadgeLink colorScheme='blue' badgeText='Read' href='https://nodejs.org/docs/latest/api/modules.html#__dirname'>Official Website</BadgeLink>
+<BadgeLink colorScheme='yellow' badgeText='Read' href='https://www.digitalocean.com/community/tutorials/nodejs-how-to-use__dirname'>Digital Ocean | how to use __dirname </BadgeLink>
```
Comment (kamranahmedse) [inline content/roadmaps/107-nodejs/content/105-nodejs-working-with-files/107-dirname.md:11]:
```suggestion
# __dirname

The `__dirname` in a node script returns the path of the folder where the current JavaScript file resides. `__filename` and `__dirname` are used to get the filename and directory name of the currently executing file.

<ResourceGroupTitle>Free Content</ResourceGroupTitle>
<BadgeLink colorScheme='blue' badgeText='Official Docs' href='https://nodejs.org/docs/latest/api/modules.html#__dirname'>Official Website</BadgeLink>
<BadgeLink colorScheme='yellow' badgeText='Read' href='https://www.digitalocean.com/community/tutorials/nodejs-how-to-use__dirname'>How to use __dirname</BadgeLink>
```

File: content/roadmaps/107-nodejs/content/105-nodejs-working-with-files/108-filename.md [hunk 9]
```diff
@@ -1 +1,7 @@
-# Filename
\ No newline at end of file
+# Filename
+
+The __filename in the Node.js returns the filename of the code which is executed. It gives the absolute path of the code file. The following approach covers how to implement __filename in the NodeJS project.
+
+<ResourceGroupTitle>Free Content</ResourceGroupTitle>
+<BadgeLink colorScheme='blue' badgeText='Read' href='https://nodejs.org/docs/latest/api/modules.html#__filename'>Official Website</BadgeLink>
+<BadgeLink colorScheme='yellow' badgeText='Read' href='https://www.geeksforgeeks.org/node-js-__filename-object/'> What is __filename </BadgeLink>
```
Comment (kamranahmedse) [inline content/roadmaps/107-nodejs/content/105-nodejs-working-with-files/108-filename.md:7]:
```suggestion
# __filename

The `__filename` in Node.js returns the filename of the executed code. It gives the absolute path of the code file. The following approach covers implementing `__filename` in the Node.js project.

<ResourceGroupTitle>Free Content</ResourceGroupTitle>
<BadgeLink colorScheme='blue' badgeText='Official Docs' href='https://nodejs.org/docs/latest/api/modules.html#__filename'>Official Docs</BadgeLink>
```

**Comment by kamranahmedse** [comment]
Please have a look at my changes and try to follow them in the future PRs.

**Comment by CodeWAdi** [comment]
Thanks for updating my PR and merging it and I'll make sure to follow how you did it in the future pr

**Pull request merged**